test/test_apps/common/test_app_base.cpp

```diff
@@ -2351,22 +2351,27 @@ std::exception sdl_exception() {
     return std::runtime_error(SDL_GetError());
 }
 
-Init device_initialization(const std::string& window_name) {
-    Init init;
-
+void device_initialization_phase_1(const std::string& window_name, Init& init)
+{
     init.window = gfxrecon::test::create_window_sdl(window_name.data(), true, 1024, 1024);
+}
 
-    gfxrecon::test::InstanceBuilder instance_builder;
-    init.instance = instance_builder.use_default_debug_messenger().request_validation_layers().build();
+void device_initialization_phase_2(InstanceBuilder const& instance_builder, Init& init)
+{
+    init.instance = instance_builder.build();
 
     init.inst_disp = init.instance.make_table();
 
     init.surface = gfxrecon::test::create_surface_sdl(init.instance, init.window);
+}
 
```

Comment:
What's the benefit of splitting this up?

Reply:
Between phase 1 and 2, there is an opportunity for test apps to customize instance creation via the InstanceBuilder.
Between phases 2 and 3, apps can customize physical device selection via the PhysicalDeviceSelector.
Between phase 3 and 4, apps can customize device creation via the DeviceBuilder.
Betwen phase 4 and 5, apps can customize swapchain creation via the SwapchainBuilder.

This is threaded together in TestAppBase::run.  However, I left open the possiblity of test apps not using TestAppBase and left the device_initialization method from VkBootstrap as an alternative.  That method does not allow for the customizations above.  I used these "phase" functions to make sure the two implementations were similar.  In the future I could see making them part of the public interface if people need customization but don't want to use TestAppBase.  Or we could force TestAppBase and collapse the functionality into TestAppBase::run.

Reply:
Don't change it for me.  Maybe as you add more features in each phase, rename them from "phase_x" to "device_initialization_<descrip>" to be clearer.